subsys/net/lib/dns/dns_cache.c

```diff
@@ -0,0 +1,174 @@
+		dns_cache_clean(cache);
+		for (size_t i = 0; i < cache->size; i++) {
+			if (!cache->entries[i].in_use) {
+				index_to_replace = i;
+				found_empty = true;
+				break;
+			} else if (cache->entries[i].creation_uptime < oldest_entry_uptime) {
+				index_to_replace = i;
+				oldest_entry_uptime = cache->entries[i].creation_uptime;
+			}
+		}
+	}
+
+	if (!found_empty) {
+		NET_DBG("OVERWRITE \"%s\"", cache->entries[index_to_replace].query);
+	}
+
+	strncpy(cache->entries[index_to_replace].query, query,
+		CONFIG_DNS_RESOLVER_MAX_QUERY_LEN - 1);
+	cache->entries[index_to_replace].data = *addrinfo;
+	cache->entries[index_to_replace].creation_uptime = k_uptime_get();
+	cache->entries[index_to_replace].ttl = ttl;
+	cache->entries[index_to_replace].in_use = true;
+	NET_DBG("Uptime: %" PRIu64, cache->entries[index_to_replace].creation_uptime);
+
+	k_mutex_unlock(cache->lock);
+
+	return 0;
+}
+
+int dns_cache_remove(struct dns_cache *cache, char const *query)
+{
+	NET_DBG("REMOVE all entries with query \"%s\"", query);
+	if (strlen(query) >= CONFIG_DNS_RESOLVER_MAX_QUERY_LEN) {
+		NET_WARN("Query string to big to be processed %u >= "
+			 "CONFIG_DNS_RESOLVER_MAX_QUERY_LEN",
+			 strlen(query));
+		return -EINVAL;
+	}
+
+	k_mutex_lock(cache->lock, K_FOREVER);
+
+	dns_cache_clean(cache);
+
+	for (size_t i = 0; i < cache->size; i++) {
+		if (cache->entries[i].in_use && strcmp(cache->entries[i].query, query) == 0) {
+			cache->entries[i].in_use = false;
+		}
+	}
+
+	k_mutex_unlock(cache->lock);
+
+	return 0;
+}
+
+int dns_cache_find(struct dns_cache const *cache, const char *query, struct dns_addrinfo *addrinfo,
+		   size_t addrinfo_array_len)
+{
+	size_t found = 0;
+
+	NET_DBG("FIND \"%s\"", query);
+	if (cache == NULL || query == NULL || addrinfo == NULL || addrinfo_array_len <= 0) {
+		return -EINVAL;
+	}
+	if (strlen(query) >= CONFIG_DNS_RESOLVER_MAX_QUERY_LEN) {
+		NET_WARN("Query string to big to be processed %u >= "
+			 "CONFIG_DNS_RESOLVER_MAX_QUERY_LEN",
+			 strlen(query));
+		return -EINVAL;
+	}
+
+	k_mutex_lock(cache->lock, K_FOREVER);
+
+	dns_cache_clean(cache);
+
+	for (size_t i = 0; i < cache->size; i++) {
+		if (!cache->entries[i].in_use) {
+			continue;
+		}
+		if (strcmp(cache->entries[i].query, query) != 0) {
+			continue;
+		}
+		if (found >= addrinfo_array_len) {
+			NET_WARN("FOUND \"%s\" but not enough space in provided buffer.", query);
+			found++;
+		} else {
+			addrinfo[found] = cache->entries[i].data;
+			found++;
+			NET_DBG("FOUND \"%s\"", query);
+		}
+	}
+
+	k_mutex_unlock(cache->lock);
+
+	if (found > addrinfo_array_len) {
+		return -ENOSR;
+	}
+
+	if (found == 0) {
+		NET_DBG("COULD NOT FIND \"%s\"", query);
+	}
+	return found;
+}
+
+/* Needs to be called when lock is already acquired */
+static void dns_cache_clean(struct dns_cache const *cache)
+{
+	for (size_t i = 0; i < cache->size; i++) {
+		if (!cache->entries[i].in_use) {
+			continue;
+		}
+		int64_t reftime = cache->entries[i].creation_uptime;
```

Comment:
About the time calculations - I wonder that whether it'd make sense, instead of storing creation time + ttl, just have a `k_timepoint_t` for the expiry time, then you could simply use `sys_timepoint_expired()` to check if the entry is still valid, instead of doing the calculations manually.

Reply:
This did however change the functionality a bit since now the entry closest to expiry is the one that will be removed instead of the oldest when additional space is needed.